wstunnel: 9.6.2 -> 9.7.0

r-vdp opened the pull request:
## Description of changes

This also removes the top-level attribute for the old Haskell-based implementation in favour of the new Rust implementation by the same author.
The Haskell-based version is still accessible from `haskellPackages`.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

r-vdp committed:
wstunnel: 9.6.2 -> 9.7.0

Aleksanaa commented:
I opened a PR https://github.com/NixOS/nixpkgs/pull/289652 but if you want to proceed just go on.

The most problematic part is the nixos module, as the new rewrite uses different command line args.

r-vdp commented:
Oh, I should've searched for that first!

Yeah, as noted there, if the old package doesn't build anymore, we can probably just hijack the name.

I will have a look at the module, to bring it up to date with the rust version. A NixOS test would've been really nice for this, maybe we can work on adding that too.

r-vdp committed:
nixos/wstunnel: update the wstunnel module to work with the new rust implementation

Co-authored-by: h7x4 <h7x4@nani.wtf>

h7x4 commented:
I think #317840 might conflict with this PR. It was just merged

r-vdp commented:
That's fine, I'll rebase on top of it

NeverBehave commented on nixos/modules/services/networking/wstunnel.nix:
```diff
@@ -1,123 +1,159 @@
       '';
-      type = types.nullOr types.str;
+      type = lib.types.nullOr (lib.types.enum [
+        "TRACE"
```
Per description, it seems enum does not cover all cases of EnvFilter, such as `warn,tokio::net=info`

r-vdp replied:
Hmm, good point, I can revert that part

r-vdp replied:
Done!

NeverBehave reviewed (approved):
LGTM if the changes has tested locally. Only checked diff.

h7x4 commented on nixos/modules/services/networking/wstunnel.nix:
```diff
@@ -228,44 +275,55 @@ let
-        PrivateDevices = true;
-        RestrictSUIDSGID = true;
-      };
+      script = with serverCfg; let
+        resolvedTlsCertificate =
+          if useACMEHost != null
+          then "${certConfig.directory}/fullchain.pem"
+          else tlsCertificate;
+        resolvedTlsKey =
+          if useACMEHost != null
+          then "${certConfig.directory}/key.pem"
+          else tlsKey;
+      in
+      ''
+        ${lib.getExe package} \
+          server \
+          ${lib.concatMapStringsSep " " (restriction: ''--restrict-to ${lib.escapeShellArg (hostPortToString restriction)}'') restrictTo} \
+          ${lib.optionalString (resolvedTlsCertificate != null) "--tls-certificate ${lib.escapeShellArg resolvedTlsCertificate}"} \
+          ${lib.optionalString (resolvedTlsKey != null) "--tls-private-key ${lib.escapeShellArg resolvedTlsKey}"} \
+          ${attrsToArgs extraArgs} \
+          ${lib.escapeShellArg "${if enableHTTPS then "wss" else "ws"}://${hostPortToString listen}"}
+      '';
```
```suggestion
      script = with serverCfg; ''
        ${lib.getExe package} server ${lib.cli.toGNUCommandLineShell { } (
          lib.recursiveUpdate
          {
            restrict-to = map hostPortToString restrictTo;
            tls-certificate = if useACMEHost != null
                              then "${certConfig.directory}/fullchain.pem"
                              else tlsCertificate;
            tls-private-key = if useACMEHost != null
                              then "${certConfig.directory}/key.pem"
                              else tlsKey;
          }
          extraAttrs
        )} ${lib.escapeShellArg "${if enableHTTPS then "wss" else "ws"}://${hostPortToString listen}"}
      '';
```

h7x4 commented on nixos/modules/services/networking/wstunnel.nix:
```diff
@@ -228,44 +275,55 @@ let
+
+      serviceConfig =
+        let
+          certConfig = config.security.acme.certs."${serverCfg.useACMEHost}";
```
```suggestion
          certConfig = config.security.acme.certs.${serverCfg.useACMEHost};
```

h7x4 commented on nixos/modules/services/networking/wstunnel.nix:
```diff
@@ -126,95 +155,113 @@ let
+          Setting this option will also enable the required `CAP_NET_ADMIN` capability
+          for the systemd service.
+        '';
+        type = lib.types.nullOr lib.types.int;
```
```suggestion
        type = lib.types.nullOr lib.types.ints.unsigned;
```

h7x4 commented on nixos/modules/services/networking/wstunnel.nix:
```diff
@@ -126,95 +155,113 @@ let
         description = "Listen on remote and forwards traffic from local. Only tcp is supported";
-        type = types.listOf (types.str);
-        default = [];
+        type = lib.types.listOf (lib.types.str);
```
```suggestion
        type = lib.types.listOf lib.types.str;
```

h7x4 commented on nixos/modules/services/networking/wstunnel.nix:
```diff
@@ -1,123 +1,152 @@
-        default = [];
+      restrictTo = lib.mkOption {
+        description = ''
+          Accepted traffic will be forwarded only to this service.
+          Set to `null` to allow forwarding to arbitrary addresses.
```
I know you have not edited this description, but could we have a look into what the original author was trying to say here? AFAIK, `null` is not an accepted value. Is arbitrary traffic accepted by default?

r-vdp replied:
The reference to null is a left-over from before we turned this option into a list, I removed that part.

Without any restrictions specified on the server, a client can indeed ask for any forwarding. Setting this option on the server, limits what a client can ask for.

h7x4 commented on nixos/modules/services/networking/wstunnel.nix:
```diff
@@ -1,123 +1,152 @@
+        Attributes of the form `argName = true;` will be translated to `--argName`,
+        and `argName = \"value\"` to `--argName value`.
+      '';
+      type = with lib.types; attrsOf (either str bool);
```
If you decide to go for the `lib.cli.toGNUCommandLineShell` thing, this can be expanded to allow for more types (I'm working on making this a `formats.cli` thing, but it's not finished yet).

```suggestion
      type = with lib.types; let
        atomType = nullOr (oneOf [ str int float bool package ]);
      in attrsOf (either atomType (listOf atomType));
```

r-vdp replied:
I will keep this one for later, since it's not directly useful currently.

r-vdp replied:
Let's do this when we actually need it, I don't think we have a use for the other types right now.

h7x4 commented on nixos/modules/services/networking/wstunnel.nix:
```diff
@@ -1,123 +1,152 @@
-    (name: value: if value == true then "--${name}" else "--${name}=${value}")
-    attrs
+
+  attrsToArgs = attrs: lib.escapeShellArgs (
+    lib.mapAttrsToList
+      (name: value: if value == true then "--${name}" else "--${name}=${value}")
+      attrs
   );
```
If you decide to go for the `lib.cli.toGNUCommandLineShell` thing, this becomes obsolete.

```suggestion
```

h7x4 commented on nixos/modules/services/networking/wstunnel.nix:
```diff
@@ -1,123 +1,152 @@
-      default = true;
-    };
+    autoStart =
+      lib.mkEnableOption "Whether this tunnel server should be started automatically." // {
```
This docstring becomes funny because of the inner working of `mkEnableOption`, [it wraps the given string in `"Whether to enable ${str}."`](https://github.com/NixOS/nixpkgs/blob/a2a0316319cc0625ab75baba5084aacb272ba6d0/lib/options.nix#L98-L105)

h7x4 commented on nixos/modules/services/networking/wstunnel.nix:
```diff
@@ -356,28 +422,43 @@ in {
+      (lib.mapAttrsToList
+        (name: serverCfg: {
+          assertion =
+            !((serverCfg.tlsCertificate != null || serverCfg.tlsKey != null) &&
+            !(serverCfg.tlsCertificate != null && serverCfg.tlsKey != null));
```
This can be reduced to xnor

```suggestion
            (serverCfg.tlsCertificate == null && serverCfg.tlsKey == null)
            || (serverCfg.tlsCertificate != null && serverCfg.tlsKey != null)
```

h7x4 commented on nixos/modules/services/networking/wstunnel.nix:
```diff
@@ -356,28 +422,43 @@ in {
+      (lib.mapAttrsToList
+        (name: serverCfg: {
+          assertion =
+            !(serverCfg.useACMEHost != null &&
+            (serverCfg.tlsCertificate != null || serverCfg.tlsKey != null));
```
The fact that a certificate needs a key is an implementation detail from the perspective of the intent behind this assertion. I'm also a bit uncertain if this assertion should be here whenever `enableHttps = false`, it might make sense to prefix the assertion with `serverCfg.enableHttps -> <pred>`

```suggestion
            !(serverCfg.useACMEHost != null && serverCfg.tlsCertificate != null);
```

h7x4 reviewed (commented):
Here's a few suggestions, although some of them are comments about things you have not authored. Please do consider them though :)

r-vdp commented on pkgs/by-name/ws/wstunnel/package.nix:
```diff
@@ -3,30 +3,46 @@
     "--skip=tcp::tests::test_proxy_connection"
   ];
 
+  doInstallCheck = true;
+  installCheckPhase = ''
+    runHook preInstallCheck
+    actual="$($out/bin/wstunnel --version)"
+    expected="${pname} ${version}"
+    echo "Check that 'wstunnel --version' returns: $expected"
+    if [[ "$actual" != "$expected" ]]; then
+      echo "'wstunnel --version' returned: $actual"
+      exit 1
+    fi
+    runHook postInstallCheck
+  '';
```
Right, good point.

r-vdp commented:
> Here's a few suggestions, although some of them are comments about things you have not authored. Please do consider them though :)

Thanks! I added most of them to this PR!

h7x4 commented on nixos/modules/services/networking/wstunnel.nix:
```diff
@@ -126,95 +148,113 @@ let
 
-      tlsVerifyCertificate = mkOption {
+      tlsVerifyCertificate = lib.mkOption {
         description = "Whether to verify the TLS certificate of the server. It might be useful to set this to `false` when working with the `tlsSNI` option.";
```
It would be very nice if we could add a note here about how this might be useful to disable when using tls with self signed certificates. Considering a lot of people might use self signed certs for this, I'd even consider disabling it by default.

h7x4 commented:
Do you mind if I push a nixos test to this PR?

h7x4 committed:
nixosTests.wstunnel: init

Co-authored-by: r-vdp <ramses@well-founded.dev>

h7x4 commented:
Would you mind adding a point to the release notes? Something like this maybe?

```
- `wstunnel` has had a major version upgrade that entailed rewriting the program in rust.
  The module was updated to accomodate for breaking changes without breaking the module API,
  but be aware that arguments listed in `services.wstunnel.{clients,servers}.<name>.extraArgs`
  might have been removed or modified upstream.
```

r-vdp committed:
nixos/wstunnel: Add a mention in the release notes

r-vdp commented:
> Would you mind adding a point to the release notes?

Yeah, I reworded it slightly since the previous PR did break the module API. I think this was hard to avoid given the upstream changes.

r-vdp commented:
I also updated the test to use the new, `runTest`-based way of defining tests, as recommended in the manual, and I changed it to use the existing fake certificates that we already have for other tests.

Aleksanaa commented on pkgs/by-name/ws/wstunnel/package.nix:
```diff
@@ -1,32 +1,44 @@
 
-rustPlatform.buildRustPackage rec {
+let
   pname = "wstunnel";
```
This is only used once

r-vdp replied:
Right, I inlined it.

JohnRTitor commented:
Result of `nixpkgs-review pr 319516` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>wstunnel</li>
  </ul>
</details>

r-vdp commented:
Are we good to merge this?

h7x4 commented:
I think so
